src/Uno.UI/Controls/NativeFramePresenter.iOS.cs

```diff
@@ -80,54 +135,241 @@ private void NativeFramePresenter_SizeChanged(object sender, SizeChangedEventArg
+								this.Log().Debug($"Pushing ViewController ({page.GetType().Name}) to replicate Frame's forward navigation.");
+							}
+							NavigationController.PushViewController(viewController, isAnimated);
+							break;
+						case NavigationMode.Refresh:
+						default:
+							// Refresh currently doesn't have an effect.
+							break;
+					}
+				}
+			}
+			else
+			{
+				// We shouldn't get here because the frame events are synchronous.
+				if (frameRequest == null)
+				{
+					this.Log().Error($"Can't process OnFrameNavigated because the request queue is empty.");
+				}
+				else
+				{
+					this.Log().Error($"Can't process OnFrameNavigated because the request in queue doesn't match the current request.");
+				}
+			}
+		}
+
+		/// <summary>
+		/// This is called on <see cref="Frame.NavigationStopped"/>.
+		/// We use this handler to remove requests cancelled by <see cref="NavigatingCancelEventArgs.Cancel"/>.
+		/// </summary>
+		private void OnFrameNavigationStopped(object sender, NavigationEventArgs e)
+		{
+			var request = new NavigationRequest(_frame, e);
+			if (TryGetFirst(_frameToControllerRequests, out var frameToControllerRequest) && NavigationRequest.Correlates(request, frameToControllerRequest))
+			{
+				if (this.Log().IsEnabled(LogLevel.Debug))
+				{
+					this.Log().Debug("Aborted navigation request because the Frame.Navigating event was cancelled.");
+				}
 
-			if (viewControllers.Length == NavigationController.ViewControllers.Length + 1 &&
-				viewControllers.Take(viewControllers.Length - 1).SequenceEqual(NavigationController.ViewControllers) &&
-				// Ensure not to call PushViewController if one has already been potentially called for the same controller and is still pending, since iOS gets all like this: "NSInvalidArgumentException Reason: Pushing the same view controller instance more than once is not supported "
-				!alreadyRequested
-			)
+				_frameToControllerRequests.RemoveFirst();
+			}
+			else
 			{
-				// Use Push/Pop when possible because they're animated more nicely by iOS
-				NavigationController.PushViewController(requestedViewController, isAnimated);
+				// We shouldn't get here because the frame events are synchronous.
+				this.Log().Error($"Can't process OnFrameNavigationStopped because the request in queue doesn't match the current request.");
 			}
-			else if (viewControllers.Length == NavigationController.ViewControllers.Length - 1 &&
-				NavigationController.ViewControllers.Take(NavigationController.ViewControllers.Length - 1).SequenceEqual(viewControllers)
-			)
+		}
+
+		/// <summary>
+		/// This is called on <see cref="Frame.BackStack"/> changed.
+		/// We use this handler to detect BackStack manipulations (like removing previous pages) and reset the <see cref="UINavigationController.ViewControllers"/> when applicable.
+		/// </summary>
+		private void OnFrameBackStackChanged(object sender, NotifyCollectionChangedEventArgs e)
+		{
+			var collection = _frame.BackStack;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove this useless assignment to local variable 'collection'.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5436574)